### PR TITLE
Set application metadata on thread during JPA server start

### LINKED ISSUE #22865: Datasource changes are not propagating to JPA during dynamic config update
Changing the `<datasource>` properties (databaseName, password, portNumber, serverName, ect) in server.xml are not reflected in JPA applications. 

During server start, JPA Container does a JNDI lookup for the DataSource and provides that DataSource to the JPA provider, EclipseLink. After the configuration update fires, EclipseLink continues using the same DataSource that JPA Container provided. The DataSourceService isn't closing the ConnectionFactory, so everything continues to work. For JPA, the application needs to be restarted in order for this to work, however the DataSourceService is not refreshing the application.

### PR DESCRIPTION
**Pull request by dazey3**
fixes #22865

It appears that the `DataSourceService` is not refreshing the application because the application metadata is not available on the thread

```
java.lang.Exception: Stack trace
	at java.base/java.lang.Thread.dumpStack(Thread.java:1383)
	at com.ibm.ws.jca.cm.AbstractConnectionFactoryService.createResource(AbstractConnectionFactoryService.java:189)
	at com.ibm.ws.resource.internal.ResourceFactoryTrackerData$1.getService(ResourceFactoryTrackerData.java:120)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse$1.run(ServiceFactoryUse.java:220)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.factoryGetService(ServiceFactoryUse.java:217)
	at org.eclipse.osgi.internal.serviceregistry.ServiceFactoryUse.getService(ServiceFactoryUse.java:118)
	at org.eclipse.osgi.internal.serviceregistry.ServiceConsumer$2.getService(ServiceConsumer.java:48)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.getService(ServiceRegistrationImpl.java:547)
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.getService(ServiceRegistry.java:533)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.getService(BundleContextImpl.java:655)
	at com.ibm.ws.jndi.internal.WSContextFactory.lambda$getService$0(WSContextFactory.java:81)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at com.ibm.ws.jndi.internal.WSContextFactory.lambda$getService$1(WSContextFactory.java:81)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at com.ibm.ws.jndi.internal.WSContextFactory.getService(WSContextFactory.java:81)
	at com.ibm.ws.jndi.internal.WSContext.getReference(WSContext.java:520)
	at com.ibm.ws.jndi.internal.WSContext.resolveObject(WSContext.java:139)
	at com.ibm.ws.jndi.internal.WSContext.lookup(WSContext.java:309)
	at com.ibm.ws.jndi.WSContextBase.lookup(WSContextBase.java:61)
	at org.apache.aries.jndi.DelegateContext.lookup(DelegateContext.java:149)
	at java.naming/javax.naming.InitialContext.lookup(InitialContext.java:409)
	at com.ibm.ws.jpa.container.osgi.internal.OSGiJPAPUnitInfo.lookupDataSource(OSGiJPAPUnitInfo.java:317)
	at com.ibm.ws.jpa.management.JPAPUnitInfo.getJPADataSource(JPAPUnitInfo.java:338)
	at com.ibm.ws.jpa.management.JPAPUnitInfo.getJtaDataSource(JPAPUnitInfo.java:385)
	at com.ibm.ws.jpa.management.JPAPUnitInfo.initialize(JPAPUnitInfo.java:765)
	at com.ibm.ws.jpa.management.JPAPxmlInfo.extractPersistenceUnits(JPAPxmlInfo.java:182)
	at com.ibm.ws.jpa.management.JPAScopeInfo.processPersistenceUnit(JPAScopeInfo.java:88)
	at com.ibm.ws.jpa.management.JPAApplInfo.addPersistenceUnits(JPAApplInfo.java:119)
	at com.ibm.ws.jpa.container.osgi.internal.JPAComponentImpl.processLibraryJarPersistenceXml(JPAComponentImpl.java:528)
	at com.ibm.ws.jpa.container.osgi.internal.JPAComponentImpl.applicationStarting(JPAComponentImpl.java:317)
	at com.ibm.ws.container.service.state.internal.ApplicationStateManager.fireStarting(ApplicationStateManager.java:51)
	at com.ibm.ws.container.service.state.internal.StateChangeServiceImpl.fireApplicationStarting(StateChangeServiceImpl.java:50)
	at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase.preDeployApp(SimpleDeployedAppInfoBase.java:547)
	at com.ibm.ws.app.manager.module.internal.SimpleDeployedAppInfoBase.installApp(SimpleDeployedAppInfoBase.java:508)
	at com.ibm.ws.app.manager.module.internal.DeployedAppInfoBase.deployApp(DeployedAppInfoBase.java:349)
	at com.ibm.ws.app.manager.ear.internal.EARApplicationHandlerImpl.install(EARApplicationHandlerImpl.java:77)
	at com.ibm.ws.app.manager.internal.statemachine.StartAction.execute(StartAction.java:161)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.enterState(ApplicationStateMachineImpl.java:1357)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.run(ApplicationStateMachineImpl.java:901)
	at com.ibm.ws.threading.internal.ExecutorServiceImpl$RunnableWrapper.run(ExecutorServiceImpl.java:245)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

Signed-off-by: Will Dazey <dazeydev.3@gmail.com>

### CODE REVIEW

**Comment by wraschke** [comment]
#libby

**Comment by dazey3** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jpa.tests.container_modifyconfig_fat

**Comment by dazey3** [comment]
#build #spawn.fullfat.buckets=com.ibm.ws.jpa.tests.container_modifyconfig_fat

**Comment by dazey3** [comment]
#run-libby-bot

File: dev/com.ibm.ws.jpa.container/src/com/ibm/ws/jpa/container/osgi/internal/JPAComponentImpl.java [hunk 1]
```diff
@@ -379,6 +391,10 @@ public void applicationStarting(ApplicationInfo appInfo) {
                 FFDCFilter.processException(e, this.getClass().getName(), "457");
             }
         } finally {
+            if (setContext) {
+                ComponentMetaDataAccessorImpl.getComponentMetaDataAccessor().endContext();
```
Comment (jgrassel) [inline dev/com.ibm.ws.jpa.container/src/com/ibm/ws/jpa/container/osgi/internal/JPAComponentImpl.java:395]: There any chance this might throw an exception?

Comment (dazey3) [reply dev/com.ibm.ws.jpa.container/src/com/ibm/ws/jpa/container/osgi/internal/JPAComponentImpl.java:395]: @jgrassel I don't believe so. Looking at the implementation, I don't see that it throws an exception. Also, there is a similar implementation in the `CDIRuntimeImpl` that I am reusing

**Comment by dazey3** [comment]
#run-libby-bot

**Review by tkburroughs — APPROVED** [review]

**Pull request merged**